Un 67 expose variables

juliandailey opened the pull request:
exposed variables
form for data input & map generation

juliandailey committed:
Merge branch 'main' of https://github.com/Calvinjk/USS-Nebula

juliandailey committed:
Merge branch 'main' of https://github.com/Calvinjk/USS-Nebula

juliandailey committed:
Revert "Merge branch 'main' of https://github.com/Calvinjk/USS-Nebula"

This reverts commit 0fa3e54f3d49b126cabbb8fcefb856bcb3d41f2b, reversing
changes made to 06f8fd3deb79ddb80aa96c1efc40de8945017b02.

juliandailey committed:
Merge branch 'main' of https://github.com/Calvinjk/USS-Nebula

juliandailey committed:
Merge branch 'main' of https://github.com/Calvinjk/USS-Nebula

juliandailey committed:
map generator updates

created forms for ease of data input
exposed variables in parameter of map generation function

Calvinjk commented on nebula-unity/Assets/scripts/DungeonMapGenerator.cs:
```diff
@@ -108,11 +109,15 @@ public class PotentialDoor {
     }
 
     // This function will generate a map given some bounding dimensions
-    public Map GenerateMap(int xSize = XDIMDEFAULT, int ySize = YDIMDEFAULT) {
+    public Map GenerateMap(int xSize, int ySize, int minDiam, int maxDiam, float sfact, int maxatt) {
         ResetGenerationVariables();
 
         mapSize.x = xSize;
         mapSize.y = ySize;
+        minRoomDiameter = minDiam;
+        maxRoomDiameter = maxDiam;
+        shapeFactor = sfact;
+        maxAttempts = maxatt;
```
Since you removed the only reference to them, you should also remove XDIMDEFAULT and YDIMDEFAULT.

....or a better idea is to use defaults for all of those variables!   If we can set those values in the inspector, we should use them as defaults!   Then we can just call MakeMeAMap() without params to keep remaking maps with the same params!

Calvinjk replied:
If we do it this way, lines 115-120 become redundant and we wont need them anymore either.  Assuming we have a function declaration something like this:

`public Map GenerateMap(int xSize = mapSize.x, int ySize = maxSize.y, int minDiam = minRoomDiameter, int maxDiam = maxRoomDiameter, float sfact = shapeFactor, int maxatt = maxAttempts)`

We can then use [keyword arguments](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments) to call GenerateMap with only the args that we want to supply.

This means we can effectively use BOTH the inspector and code to efficiently call the function.  So we can set default values in the inspector, then change only ones we want to change in code, such as:

`GenerateMap(sfact: 8, maxatt: 1000)`

juliandailey committed:
update

fixed infinite loop
added sentinels for -1 default values for each field

Calvinjk commented on nebula-unity/Assets/scripts/DungeonMapGenerator.cs:
```diff
@@ -118,6 +119,21 @@ public class PotentialDoor {
         maxRoomDiameter = maxDiam;
         shapeFactor = sfact;
         maxAttempts = maxatt;
+=======
```
Looks like you have some artifacts here leftover from a stash. (Lines 112, 122, and 136)

Calvinjk replied:
Actually there are a lot more in GameManager.cs

Calvinjk commented on nebula-unity/Assets/scripts/DungeonMapGenerator.cs:
```diff
@@ -209,6 +225,28 @@ public class PotentialDoor {
         return mapScript;
     }
 
+    int[] ReplaceSentinels(int xSize, int ySize, int minDiam, int maxDiam, int maxatt){
+        int[] sentsreturn = new int[5];
+        if (xSize == -1)
+            sentsreturn[0] = XDIMDEFAULT;
+        if (ySize == -1)
+            sentsreturn[1] = YDIMDEFAULT;
+        if (minDiam == -1)
+            sentsreturn[2] = minRoomDiameter;
+        if (maxDiam == -1)
+            sentsreturn[3] = maxRoomDiameter;
+        if (maxatt == -1)
+            sentsreturn[4] = maxAttempts;
+        return sentsreturn;
+    }
+
+    float ReplaceSentinelsFloat(float sfact){
+        if (sfact == -1f)
+            return shapeFactor;
+        else
+            return sfact;
+    }
+
```
Your problem likely lies here.  ReplaceSentinelsFloat() looks good, it replaces the sentinel value if it exists, otherwise it returns the original value.    However let's take a look at ReplaceSentinels().  If `maxatt != -1`, what happens?  What does maxatt get set to on line 135?  What should it be?

Calvinjk replied:
Msg me once you figure it out but before you make the change, as I have a good solution for this.

juliandailey committed:
update 2

fused sentinel functions into one with pass by reference variables

Calvinjk commented on nebula-unity/Assets/scripts/GameManager.cs:
```diff
@@ -34,27 +34,17 @@ public class GameManager : MonoBehaviour {
         else
             minRoomDiameter = int.Parse(GameObject.Find("MinRoomDiameter").GetComponent<Text>().text);
         if (minRoomDiameter < 3)
             minRoomDiameter = 3;
```
I'd remove the `|| minRoomDiameter == 3` check from line 43.  I believe a room diameter of 3 is allowed last we talked.  But if we want to move the minimum, this is easy enough to do with a swap to line 48 instead

Calvinjk committed:
Set mapSize

Calvinjk commented on nebula-unity/Assets/scripts/DungeonMapGenerator.cs:
```diff
@@ -120,6 +120,7 @@ public class PotentialDoor {
         map.transform.position = new Vector3(0f, 0f, 0f);
         Map mapScript = map.AddComponent<Map>();
         mapScript.SetMapSize(xSize, ySize);
+        mapSize = new Vector2Int(xSize, ySize);
```
I just put this here to show you that all you need to do to fix your issue was set mapSize.  You can actually do this however you want.

juliandailey committed:
final changes

changing maxatt
changing logic of min > max

Calvinjk commented on nebula-unity/Assets/scripts/DungeonMapGenerator.cs:
```diff
@@ -109,11 +109,11 @@ public class PotentialDoor {
 
     // This function will generate a map given some bounding dimensions
-    public Map GenerateMap(int xSize = -1, int ySize = -1, int minDiam = -1, int maxDiam = -1, float sfact = -1f, int maxatt = -1) {
+    public Map GenerateMap(int xSize = -1, int ySize = -1, int minDiam = -1, int maxDiam = -1, float sfact = -1f, int maxatmp = -1) {
```
Can we camel-case?    `maxatmp` -> `maxAtmp`

Calvinjk commented on nebula-unity/Assets/scripts/GameManager.cs:
```diff
@@ -18,9 +18,81 @@ public class GameManager : MonoBehaviour {
+        else
+            minRoomDiameter = int.Parse(GameObject.Find("MinRoomDiameter").GetComponent<Text>().text);
+        if (minRoomDiameter < 3)
+            minRoomDiameter = 3;
```
I think the `|| minRoomDiameter == 3` is definitely unnecessary.  The line before (38) sets `minRoomDiameter` to 1, so this will never be true.

Calvinjk replied:
Similar to my comment below, we should avoid hard-coded values.  I'd appreciate a top-level var such as `public static int MIN_ALLOWED_ROOM_DIAMETER = 3;` that we can set at the beginning of a file instead of searching the code for this line.

Calvinjk commented on nebula-unity/Assets/scripts/GameManager.cs:
```diff
@@ -18,9 +18,81 @@ public class GameManager : MonoBehaviour {
+        else
+            maxRoomDiameter = int.Parse(GameObject.Find("MaxRoomDiameter").GetComponent<Text>().text);
+        if (maxRoomDiameter > 20)
+            maxRoomDiameter = 20;
```
For now I would actually remove these lines.  I don't want to specify an upper limit quite yet, and especially not a hardcoded value.  If you want to leave in an upper limit, we could just swap it to an inspector-level variable.

Calvinjk commented on nebula-unity/Assets/scripts/GameManager.cs:
```diff
@@ -18,9 +18,81 @@ public class GameManager : MonoBehaviour {
+        if (shapeFactor > 10f)
+            shapeFactor = 10f;
+		return shapeFactor;
+    }
```
Just to put down in words what I said during the meeting, I am planning to change ShapeFactor to allow the range [-10, 10].  So here we will need to change the sentinel and include a minimum check as well.

Calvinjk commented on nebula-unity/Assets/scripts/GameManager.cs:
```diff
@@ -18,9 +18,81 @@ public class GameManager : MonoBehaviour {
+            shapeFactor = 10f;
+		return shapeFactor;
+    }
+    int SetMaxAttempts(){
+		int maxAttempts;
+        if (GameObject.Find("MaxAttempts").GetComponent<Text>().text == "")
+            maxAttempts = -1;
+        else
+            maxAttempts = int.Parse(GameObject.Find("MaxAttempts").GetComponent<Text>().text);
+		return maxAttempts;
+    }
```
Since you are doing some input validation in the other functions, I think we should do some here as well.   maxAttempts should be any integer greater than 0.

Calvinjk commented on nebula-unity/Assets/scripts/GameManager.cs:
```diff
@@ -18,9 +18,81 @@ public class GameManager : MonoBehaviour {
+    }
+    int SetSize(){
+        int tempsize;
+        if (GameObject.Find("Size").GetComponent<Text>().text != "")
+            tempsize = int.Parse(GameObject.Find("Size").GetComponent<Text>().text);
```
For the sake of consistency, could you swap this check to `if ==` and not `if !=`?  Just to match the checks in the other functions.

Calvinjk commented on nebula-unity/Assets/scripts/GameManager.cs:
```diff
@@ -18,9 +18,81 @@ public class GameManager : MonoBehaviour {
+            tempsize = int.Parse(GameObject.Find("Size").GetComponent<Text>().text);
+        else
+            tempsize = -1;
+        if (tempsize < 50)
+            tempsize = 50;
```
Just as above, lets stay away from hardcoded values!   Top-level var such as `public static int MIN_MAP_DIMENSION = 50;` would be nicer.   Additionally, if you are willing to do the extra work, I'd like to allow non-square maps.   So letting us specify an x and y separately (such as 100x500 to create a long, thin map) would be really nice.

Calvinjk commented on nebula-unity/Assets/scripts/GameManager.cs:
```diff
@@ -18,9 +18,81 @@ public class GameManager : MonoBehaviour {
+    }
+
+    public Map MakeMeAMap(){
+        int[] diameterrange = SetRoomDiameterRange();
+        float sfactor = SetShapeFactor();
+        int maxattempts = SetMaxAttempts();
+        int returnsize = SetSize();
+
+        return generatorScript.GenerateMap(xSize: SetSize(), ySize: SetSize(), maxDiam: SetRoomDiameterRange()[1], minDiam: SetRoomDiameterRange()[0], sfact: SetShapeFactor(), maxatmp: SetMaxAttempts());
```
These methods are much closer in function to getters than setters.  I move we swap the names to Get____().   

In addition, you are actually calling all these methods twice (thrice in the case of SetRoomDiameterRange()).  We really only need to call each once, so either use the vars you saved or call them directly in the GenerateMap() call, not both.

Calvinjk reviewed (changes requested):
I am not sure why they keep getting added, but all of the `nebula-unity/Library` files should not be added.  Could you remove them from this PR and also check your local .gitignore (located in the `nebula-unity` folder) and make sure it matches the one on master?

juliandailey committed:
camelCase

camelCase

juliandailey committed:
final changes but for real this time

- added static variables instead of hard coding
- added separate x y size
- added bottom check on shape factor

juliandailey committed:
removed library files

removed library files

Calvinjk commented:
Final request I promise!   You have code of this pattern quite frequently:
```
if (GameObject.Find("MaxAttempts").GetComponent<Text>().text == "")
    maxAttempts = -1;
else
    maxAttempts = int.Parse(GameObject.Find("MaxAttempts").GetComponent<Text>().text);
```

GameObject.Find() is a pretty resource-intensive call, as it has to iterate through EVERY GameObject in scene until it finds the one with a matching name.  As our scenes get larger, this call gets slower and slower.   While this particular instance isn't much of a danger (the objects are in the menu which does not have many GameObjects), I believe in building good habits!   In your code above, GameObject.Find() is called twice, but we could reduce that down to one by storing the found GameObject:

```
Text blah = GameObject.Find("MaxAttempts").GetComponent<Text>().text;
if (blah == "")
    maxAttempts = -1;
else
    maxAttempts = int.Parse(blah);
```

This makes the code a bit more readable (and also reduces extra calls to GetComponent() as well!

juliandailey committed:
text update

reduced calls to gameobject.find
